File: usaco2025jan/2.py
def solve():
    import sys
    input = sys.stdin.read
    data = input().split()
    
    T = int(data[0])  # Number of test cases
    idx = 1
    results = []
    
    for _ in range(T):
        N, M = map(int, data[idx:idx+2])
        idx += 2
        a = list(map(int, data[idx:idx+N]))
        idx += N
        
        # Compute residues modulo M
        residues = [x % M for x in a]
        
        # Count frequency of each residue
        from collections import Counter
        freq = Counter(residues)
        
        # Sort unique residues
        unique_residues = sorted(freq.keys())
        prefix_sum = [0] * len(unique_residues)
        total_count = 0
        
        # Compute prefix sums for fast alignment cost calculation
        for i, r in enumerate(unique_residues):
            total_count += freq[r]
            prefix_sum[i] = total_count
        
        # Minimize operations
        min_cost = float('inf')
        for i, target_r in enumerate(unique_residues):
            cost = 0
            
            # Calculate cost to align all residues to target_r
            for r, count in freq.items():
                diff = abs(r - target_r)
                cost += min(diff, M - diff) * count
            
            min_cost = min(min_cost, cost)
        
        results.append(min_cost)
    
    # Print all results
    sys.stdout.write("\n".join(map(str, results)) + "\n")

solve()

File: usaco2025jan/2.cpp
#include <bits/stdc++.h>
using namespace std;

void solve() {
    int T; // Number of test cases
    cin >> T;
    while (T--) {
        long long N, M; // Size of array and modulo value
        cin >> N >> M;
        vector<long long> a(N);

        // Read the array
        for (int i = 0; i < N; i++) {
            cin >> a[i];
        }

        // Frequency map for residues modulo M
        vector<long long> freq(M, 0);
        for (int i = 0; i < N; i++) {
            freq[a[i] % M]++;
        }

        // Calculate the minimal cost
        long long min_cost = LLONG_MAX;

        // Iterate over all possible residues x % M
        for (long long x = 0; x < M; x++) {
            long long cost = 0;
            for (long long r = 0; r < M; r++) {
                long long delta = abs(x - r); // Distance between residues
                delta = min(delta, M - delta); // Circular distance modulo M
                cost += delta * freq[r]; // Multiply by frequency of residue
            }
            min_cost = min(min_cost, cost); // Track minimum cost
        }

        // Output the minimum cost
        cout << min_cost << endl;
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    solve();
    return 0;
}
